### PR TITLE
[MDEP-579] - Add mirrors, proxies and authentication for remote repositories.

### PR DESCRIPTION
**Pull request by pmoerenhout**
- [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

This new pull request makes pull request https://github.com/apache/maven-dependency-plugin/pull/18 redundant.

It includes now a Jetty server on port 8000 to test the authentication.

### CODE REVIEW

File: src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java [hunk 1]
```diff
@@ -96,6 +117,27 @@ public void testRemoteRepositories()
...
+        server.start();
+
+        setVariableValueToObject( mojo, "remoteRepositories", "myserver::default::http://localhost:8000/maven" );
+        mojo.setGroupId( "test" );
+        mojo.setArtifactId( "test" );
+        mojo.setVersion( "1.0" );
+
+        mojo.execute();
+
+        server.stop();
```
Comment (Tibor17) [inline src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java]: Pls call it within the `finally` block. This is a typical problem of integration tests that they failed and the server is never stopped.

File: src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java [hunk 2]
```diff
@@ -96,6 +117,27 @@ public void testRemoteRepositories()
...
 
+    /**
+     * Test remote repositories parameter with basic authentication
+     *
+     * @throws Exception in case of errors
+     */
+    public void testRemoteRepositoriesAuthentication()
+        throws Exception
+    {
+        org.eclipse.jetty.server.Server server = createServer( 8000 );
```
Comment (Tibor17) [inline src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java]: A concrete prort cannot be. We use the wildcard port `0.0.0.0` and the client connects to the server's local port.

File: src/test/resources/unit/get-test/repository/test/test/1.0/test-1.0.pom [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>test</groupId>
+  <artifactId>test</artifactId>
+  <version>1.0</version>
+
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-compiler-plugin</artifactId>
```
Comment (Tibor17) [inline src/test/resources/unit/get-test/repository/test/test/1.0/test-1.0.pom]: We don't need to have a plugin for this. This is enough:
```
  <properties>
    <maven.compiler.source>1.7</maven.compiler.source>
    <maven.compiler.target>1.7</maven.compiler.target>
  </properties>
```

File: src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java [hunk 4]
```diff
@@ -48,7 +62,14 @@ protected void setUp()
...
 
         LegacySupport legacySupport = lookup( LegacySupport.class );
-        legacySupport.setSession( newMavenSession( new MavenProjectStub() ) );
+        MavenSession session = newMavenSession(new MavenProjectStub());
+        Settings settings = session.getSettings();
+        Server server = new Server();
+        server.setId( "myserver" );
+        server.setUsername( "foo" );
+        server.setPassword( "bar" );
+        settings.addServer(server);
```
Comment (Tibor17) [inline src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java]: add spaces like his: `settings.addServer( server )`

File: src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java [hunk 5]
```diff
@@ -150,4 +192,47 @@ public void testParseRepository()
             // expected
         }
     }
+
+    private ContextHandler createContextHandler() {
```
Comment (Tibor17) [inline src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java]: The same here (spaces).
Bracket should be on a new line:
```
private ContextHandler createContextHandler()
{
```

File: src/test/resources/unit/get-test/realm.properties [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# username: password[,rolename ...]
+foo: bar,userrole
```
Comment (Tibor17) [inline src/test/resources/unit/get-test/realm.properties]: add an emty line.

File: src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java (hunk 1 above)
Comment (pmoerenhout) [reply src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java]: I've added the server.stop() to a finally block.

File: src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java (hunk 2 above)
Comment (pmoerenhout) [reply src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java]: I start the server now with a random port, and use server.getURI() to generate the remoteRepositories parameter.

File: src/test/resources/unit/get-test/repository/test/test/1.0/test-1.0.pom (hunk 3 above)
Comment (pmoerenhout) [reply src/test/resources/unit/get-test/repository/test/test/1.0/test-1.0.pom]: Done. Thanks!

File: src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java (hunk 4 above)
Comment (pmoerenhout) [reply src/test/java/org/apache/maven/plugins/dependency/TestGetMojo.java]: Done.

File: src/test/resources/unit/get-test/realm.properties (hunk 6 above)
Comment (pmoerenhout) [reply src/test/resources/unit/get-test/realm.properties]: Empty line added!

**Comment by rfscholte** [comment]
Fixed in https://github.com/apache/maven-dependency-plugin/commit/e897911067c643ee891bfc2b05a85f8c7ca658d5

**Comment by jira-importer** [comment]
Resolve #853

**Pull request closed without merging**